File: Mathematical Algorithms/Moser-de Bruijn Sequence/MoserDeBruijn.py
# Function to generate nth term of Moser-de Bruijn Sequence
def gen(n):
    S = [0, 1]
    for i in range(2, n + 1):
        # S(2 * n) = 4 * S(n)
        if i % 2 == 0:
            S.append(4 * S[int(i / 2)])
        # S(2 * n + 1) = 4 * S(n) + 1
        else:
            S.append(4 * S[int(i / 2)] + 1)
    z = S[n]
    return z


# Generating the first 'n' terms of Moser-de Bruijn Sequence
def moserDeBruijn(n):
    for i in range(n):
        print(gen(i), end=" ")

File: Mathematical Algorithms/Moser-de Bruijn Sequence/MoserDeBruijn.java
public class MoserDeBruijn {
    // Function to generate nth term of Moser-de Bruijn Sequence
    static int gen(int n) {
        int[] S = new int[n + 1];
        S[0] = 0;
        if (n != 0)
            S[1] = 1;
        for (int i = 2; i <= n; i++) {
            // S(2 * n) = 4 * S(n)
            if (i % 2 == 0)
                S[i] = 4 * S[i / 2];
                // S(2 * n + 1) = 4 * S(n) + 1
            else
                S[i] = 4 * S[i / 2] + 1;
        }
        return S[n];
    }

    // Generating the first 'n' terms of Moser-de Bruijn Sequence
    static void moserDeBruijn(int n) {
        for (int i = 0; i < n; i++)
            System.out.print(gen(i) + " ");
    }
}
